### ISSUE TITLE
Periodic blackening of textures with Mesa / radeonsi

### ISSUE THREAD

**Issue by shmerl** [virtual-programming/witcher2-linux#175]
When running Witcher 2 using Mesa 13.0.2 and radeonsi (RX 480), at times textures turn black with kind of transition through black mesh). Seems some antialiasing issues?

I'm running it on max settings, with ubersampling disabled. See attached video for example of this behavior (play it in a loop using mpv for example).
[witcher2_black_texture.zip](https://github.com/virtual-programming/witcher2-linux/files/644239/witcher2_black_texture.zip)

**Comment by vpltd-peterm** [comment]
We have seen the issue you mention for quite a while now, and we couldnt find a direct cause. As it never happens on the nvidia or fglrx (or amdgpu-pro) drivers, we believe its a mesa bug.

I think we have mentioned it to mesa developers, but not officially... so please go ahead and report it :)

**Comment by shmerl** [comment]
It looks like it was already reported here: https://bugs.freedesktop.org/show_bug.cgi?id=98238

**Comment by vpltd-peterm** [comment]
OK. We have not seen this behavior on the Mac either (which uses practically the same D3D9->GL engine) so we can only assume it is a Mesa bug.

**Comment by licaon-kter** [comment]
https://lists.freedesktop.org/archives/mesa-dev/2017-January/139432.html ?

**Comment by shmerl** [comment]
Interesting. This looks like a workaround and to use it would require to create some hook I suppose for  si_llvm_emit_fetch? Or to have a whole custom Mesa build.

From the comment there:

    /* If this is 1, preload FS inputs at the beginning of shaders. Otherwise,
     * reload them at each use.
     *
     * This must be 1 for Witcher 2 to render correctly. The cause of the Witcher 2
     * issue is still unknown. I only know that M0 is correct throughout the whole
     * shader.
     */
    #define SI_PRELOAD_FS_INPUTS 1

**Comment by edmondo** [comment]
I applied the patches from Marek. They fix the problem for me.

**Comment by vpltd-peterm** [comment]
I recall speaking to Marek about this problem once, and we agreed it was a Mesa bug, though he wasnt sure why. I don't agree that its a "workaround" for our game, because Mesa is the only driver this happens on.

**Comment by edmondo** [comment]
From Marek's commentary, I understand it as a workaround in mesa, allowing the driver to render correctly in Witcher 2. Why the workaround is needed is unknown.

Anyways, good news is: the game now looks nicer than before :)

**Comment by vpltd-peterm** [comment]
Yes, it's all good otherwise :) Im glad this has finally been solved, as it's pretty much the only sore point on using Mesa with Witcher 2. It performs much better than fglrx does thats for sure!

**Comment by shmerl** [comment]
Some report that this fix is [only partial](https://bugs.freedesktop.org/show_bug.cgi?id=98238#c14).
